R/run-nmtran.R

```diff
@@ -0,0 +1,246 @@
+#'
+#' # Generate and return `FDATA`
+#' fdata <- nm_fdata(mod)
+#'
+#' }
+#' @name nmtran
+NULL
+
+
+#' @describeIn nmtran Run `NM-TRAN` on a model object to validate its control
+#' stream for correct coding before submission.
+#' @export
+run_nmtran <- function(
+    .mod,
+    .config_path = NULL,
+    nmtran_exe = NULL,
+    delete_on_exit = TRUE
+){
+  check_model_object(.mod, "bbi_nonmem_model")
+
+  # Capture NONMEM and NMTRAN options
+  nmtran_exe <- locate_nmtran(.mod, .config_path, nmtran_exe)
+  nm_ver <- attr(nmtran_exe, "nonmem_version")
+
+  mod_path <- get_model_path(.mod)
+  data_path <- get_data_path_from_ctl(.mod)
+
+  # make temporary directory in current directory
+  mod_name <- fs::path_ext_remove(basename(mod_path))
+  temp_folder <- paste0("nmtran_", mod_name, "_", basename(tempdir()))
+  if(fs::dir_exists(temp_folder)) fs::dir_delete(temp_folder)
+  fs::dir_create(temp_folder)
+  if(isTRUE(delete_on_exit)){
+    on.exit(unlink(temp_folder, recursive = TRUE, force = TRUE))
+  }
+
+  # Copy model
+  file.copy(mod_path, temp_folder, overwrite = TRUE)
+  nmtran_mod <- new_model(file.path(temp_folder, basename(mod_path)), .overwrite = TRUE)
+
+  # Copy dataset & overwrite $DATA record of new model
+  # NMTRAN will error if data cannot be found
+  if(fs::file_exists(data_path)){
+    file.copy(data_path, temp_folder, overwrite = TRUE)
+    # overwrite $DATA record of new model
+    modify_data_path_ctl(nmtran_mod, data_path = basename(data_path))
+  }
+
+  # Run NMTRAN
+  nmtran_results <- c(
+    list(
+      nmtran_exe = as.character(nmtran_exe),
+      nonmem_version = nm_ver,
+      absolute_model_path = mod_path
+    ),
+    execute_nmtran(nmtran_exe, mod_path = basename(mod_path), dir = temp_folder)
+  )
+
+  # assign class and return
+  class(nmtran_results) <- c(NMTRAN_PROCESS_CLASS, class(nmtran_results))
+  return(nmtran_results)
+}
+
+
+#' @describeIn nmtran Executes `run_nmtran` on a `bbi_nonmem_model` and
+#'  returns the `NM-TRAN` dataset (`FDATA`)
+#' @export
+nm_fdata <- function(
+    .mod,
+    .config_path = NULL,
+    nmtran_exe = NULL
+){
+  nmtran_p <- run_nmtran(.mod, .config_path, nmtran_exe, delete_on_exit = FALSE)
+  on.exit(fs::dir_delete(nmtran_p$run_dir))
+
+  if(nmtran_p$status_val != 0){
+    # trim output
+    output_lines <- nmtran_p$output_lines[!grepl("^\\s+$", nmtran_p$output_lines)]
+    rlang::warn(
+      c(
+        "NM-TRAN was unsuccessful and returned the following messages:",
+        paste(output_lines, collapse = "\n")
+      )
+    )
+  }
+
+  # Attempt to read in FDATA (even if status_val is not 0)
+  #  - FDATA can still be read in in _some scenarios_ where NM-TRAN fails
+  fdata_path <- file.path(nmtran_p$run_dir, "FDATA")
+  if(fs::file_exists(fdata_path)){
+    input_data <- nm_data(.mod) %>% suppressMessages()
+    fdata <- tryCatch({
+      nm_file_impl(fdata_path, skip = 0) %>%
+        stats::setNames(names(input_data))
```

Comment:
Yeah i've actually been working on this since late last week. It's a bit involved, but im currently trying to account for dropped columns not specifically defined to be dropped/skipped.

We're not in a huge rush to get this PR merged though, and the FDATA aspect is a key feature of this. I would opt to figure this aspect out here and delay merging given my current findings.

> Also, the approach for reading the file in doesn't account for multi-line records in FDATA, right?

No, but was planning on revisiting this after I addressed first the column names, and then some of your older feedback surrounding the nmtran call.

> I wondered whether it'd be better to require NONMEM 7.5 for this functionality and use FDATA.csv instead

I thought about this too. Would make things more convenient, but was hoping to avoid that limitation

Reply:
@kylebaron do you have any thoughts about the `NOWIDE` vs `WIDE` aspect? I hadnt thought about this too hard before but it may actually be a problem with how I was planning on addressing the column names. `PRED_IGNORE_DATA` may also need some handling, but think it could be ok to just error out for that.

Reply:
> We're not in a huge rush to get this PR merged though, and the FDATA aspect is a key feature of this

 AFAICT they're distinct features, and the NMTRAN check has value outside of returning FDATA.  As a reviewer, I would very much prefer we focus on get the check into a solid state before dealing with the FDATA complication.

Also, it'd be helpful if you'd post a summary of the current state and what you're revisiting.  Right now, this is marked as ready for review both here and in the internal tracker, but you've now told me in two spots that things are incomplete and changes are inflight.

Reply:
Going to mark this as resolved since was moved to a separate PR. Made a note in one of the commit messages that this still needs to be addressed there.